### ISSUE TITLE
Support a #[serde(validate = "some_function")] attribute on fields

### ISSUE THREAD

**Issue by sfackler** [serde-rs/serde#939]
I occasionally run into fields in a serde-deserializable type that require some extra validation after deserialization, for example "a u32 >= 2". The way to do this now is to make a custom deserialize function that deserializes and then does the check:

```rust
#[derive(Deserialize)]
struct Thing {
    #[serde(deserialize_with = "validate_foo")]
    foo: u32,
}

fn validate_foo<'de, D>(d: D) -> Result<u32, D::Error>
    where D: de::Deserializer<'de>
{

    let value = u32::deserialize(d)?;

    if value < 2 {
        return Err(de::Error::invalid_value(de::Unexpected::Unsigned(value as u64),
                                            &"a value at least 2"));
    }

    Ok(value)
}
```

However, it'd be nice if it were possible to specifically tell serde to use the standard deserializer but apply a validation function after the field is deserialized:

```rust
#[derive(Deserialize)]
struct Thing {
    #[serde(validate = "validate_foo")]
    foo: u32,
}

fn validate_foo<E>(v: &u32) -> Result<(), E>
    where E: de::Error
{
    if value < 2 {
        Err(de::Error::invalid_value(de::Unexpected::Unsigned(value as u64),
                                     &"a value at least 2"));
    } else {
        Ok(())
    }
}
```

**Comment by Keats** [comment]
That's very similar to one of my crate (https://github.com/Keats/validator) with a wip new version at https://github.com/Keats/validator/pull/27

**Comment by lucab** [comment]
I guess this is a duplicate of https://github.com/serde-rs/serde/issues/642.

**Comment by dtolnay** [comment]
I don't think this is a duplicated of #642. This is for a field attribute and that one is for a container attribute. They are useful in different situations.

**Comment by TedDriggs** [comment]
@sfackler, how would you feel about this:

```rust
#[derive(Deserialize)]
struct Thing {
    #[serde(and_then = "validate_foo")]
    foo: u32,
}

fn validate_foo<E>(v: u32) -> Result<u32, E>
    where E: de::Error
{
    if value < 2 {
        Err(de::Error::invalid_value(de::Unexpected::Unsigned(value as u64),
                                     &"a value at least 2"));
    } else {
        Ok(value)
    }
}
```

That would enable both validation and post-deserialize transformation, which is something I've found myself reaching for in the past.

**Comment by Mingun** [comment]
I think this one is superseded by #1550 and can be closed in preference to it

**Comment by avkonst** [comment]
I think this issue should be closed as it seems there is a solution with `and_then`. Otherwise, it makes the impression it requires work....

**Comment by TedDriggs** [comment]
@avkonst it appears that #1858 was closed without being merged, meaning this particular issue probably will need to instead reference a documentation update.

@dtolnay your logic for individual fields in [this comment](https://github.com/serde-rs/serde/pull/1858#pullrequestreview-575089757) makes sense; the place where I've found myself reaching for `and_then` is with complex objects such as the [OASIS CACAO specification](https://docs.oasis-open.org/cacao/security-playbooks/v1.0/csd03/security-playbooks-v1.0-csd03.html#_28n7liccql2x). That has requirements such as:

1. If both `valid_from` and `valid_until` are specified, then `valid_from` MUST be less than `valid_until`
2. All variables mentioned in a workflow step MUST be defined either in the playbook's variables or in the step's variables

If I'm also providing some other avenue for making these objects, then I likely already have a function with the signature `validate(Playbook) -> Result<Playbook, AnErrSerdeCanWorkWith>` in existence, and I'd like to have `serde` run the struct through that.

An alternative would be a custom `Deserialize` impl on the `Playbook` struct, like this:

```rust
pub struct Playbook {
    // fields
}

impl<'de> Deserialize<'de> for Playbook {
    fn deserialize<D>(deserializer: D) -> Result<Self, D::Error>
    where
        D: serde::Deserializer<'de>,
    {
        #[derive(Deserialize)]
        struct UncheckedPlaybook {
            // fields...
        }

        impl TryFrom<UncheckedPlaybook> for Playbook {
            type Error = AnErrorThatImplsDisplay;

            fn try_from(_: UncheckedPlaybook) -> Result<Self, Self::Error> {
                // elided...
            }
        }

        UncheckedPlaybook::deserialize(deserializer)?.try_into().map_err(D::Error::custom)
    }
}
```

That _works_, but when the struct in question has a lot of fields that's code drift waiting to happen. I've considered making a crate to solve this that would allow the generation of a replica struct with different `serde` attributes, but that seems like overkill for the common case.

**Comment by AsafFisher** [comment]
> ```rust
> elided
> ```

But this check is not lazy! Meaning, it will be performed only when you finish to serialize `UncheckedPlaybook`.  One might want a check to accrue in the middle of the serialization. For example, `ELF` magic for elf binaries or checksum.
```rust
pub struct Data{
data: T
}
```
My serialized Data object has a `MAGIC`, `md5sum` and a `total_size` virtual fields before the data field.
I want my decerializer to check the `MAGIC` and the `md5sum` before it serializes anything.

**Comment by rapiz1** [comment]
@dtolnay 's comment https://github.com/serde-rs/serde/pull/1858#pullrequestreview-575089757 provides a quick example of validation using `deserialize_with`. I feel the need to link it here to help those who's looking for a solution.

Maybe this should be summed up and added to the documentation.

**Comment by schneems** [comment]
> Maybe this should be summed up and added to the documentation.

Yes, I agree. I feel there are several issues and prs that we could close out as "good enough" with some existing solutions documented.

**Comment by CobaltCause** [comment]
I just realized that [one can use const generics for integer validators](https://play.rust-lang.org/?version=stable&mode=debug&edition=2021&gist=7a55acf2593ac238f125dd10e24802d5); a bit much to declare but pretty nice to use.

**Comment by sickill** [comment]
I feel like conflating parsing (aka deserialization, what serde was built for) with value validation is not the right way to go. I know it's tempting to combine the two operations, it'd be convenient, but convenience invites complexity. Validation could easily be done as a second step, on the serde deserialized data structure. @Keats's validator crate seems nice (my understanding is you call `.validate()` on the struct after serde successfully deserialized).

**Comment by ssokolow** [comment]
Personally, I think that what people are asking for as "validation" is more a subset of parsing.

Parsing a sequence of UTF-8 bytes into an XML DOM doesn't become "validation" just because `<` or `>` or `&` are only valid in certain positions, and an argument can be made that one shouldn't need to newtype *everything* just to enforce that an integer named `week_of_year` only has a certain valid range.

...that said, my approach generally *is* to newtype everything since that both lets me push validation into Serde and allows the compiler to prevent conflation of values.

**Comment by CobaltCause** [comment]
<sup>*Parse, don't validate* is probably one of my top 3 favorite blog posts of all time</sup>

Another reason I favor newtypes over a second "validation" step is that without newtypes you can write this code:

```rust
let x = Stuff {
    between_1_and_3: 200,
};
```

which is clearly wrong. But with my example above, you can do either of:

```rust
between_1_and_3: BoundedUsize::clamp(200),
```

```rust
between_1_and_3: BoundedUsize::new(200).expect("out of bounds"),
```

And you could feasibly write a macro like this that would cause a compile error instead of a runtime one:

```
between_1_and_3: bounded_usize!(200),
```

I think the common name for this pattern is "correct by construction", which seems more in line with the way the rest of Rust works.

**Comment by kana-rus** [comment]
"Parse, don't validate" is ideal in many situations, but if you need _serde with validation_, check out SerdeV ( https://crates.io/crates/serdev ), essentially serde with #[serde(validate = "some_function")] container attribute. This allows you to validate automatically in deserialzing with no boilerplate.

**Comment by Binero** [comment]
> _Parse, don't validate_ is probably one of my top 3 favorite blog posts of all time
> 
> Another reason I favor newtypes over a second "validation" step is that without newtypes you can write this code:
> 
> let x = Stuff {
>     between_1_and_3: 200,
> };
> 
> which is clearly wrong. But with my example above, you can do either of:
> 
> between_1_and_3: BoundedUsize::clamp(200),
> 
> between_1_and_3: BoundedUsize::new(200).expect("out of bounds"),
> 
> And you could feasibly write a macro like this that would cause a compile error instead of a runtime one:
> 
> ```
> between_1_and_3: bounded_usize!(200),
> ```
> 
> I think the common name for this pattern is "correct by construction", which seems more in line with the way the rest of Rust works.

Sorry for getting back to such an old post, but this sort-of begs the question. Sure, for `Stuff`, validation wouldn't be too useful, but what if we were to implement `Deserialize` for `BoundedUsize`? Sure, we can have a custom `deserialize_with`, but the issue merely suggests an alternative way to achieve the same.

**Comment by CobaltCause** [comment]
> what if we were to implement `Deserialize` for `BoundedUsize`?

What I would do here is either use `deserialize_with` or write a manual `Deserialize` impl, either of which would be implemented in terms of the public/checked/safe interface for creating values of the type. There's an example of this in the playground link I posted in a previous comment. I'm not personally convinced that `validate` adds much value in this situation.

---

In general I'd also be worried that `validate` would guide people away from defining newtypes. It seems like the motivating example for this issue was trimmed down from a practical case where you'd have a struct with several fields, some of which you'd want special validation logic for, and a `validate` attribute would probably be the path of least lines of code to achieve that. I think this kind of usage would make the resulting code harder to maintain in the long term, for a few reasons:

* Defining a newtype for each thing you need validation logic for makes it easier to reuse that validation logic across types containing that thing.
* Defining a newtype and using it across the codebase makes it easier to push validation up the stack so that the validation can be done earlier.
* Not defining a newtype allows people to make the mistake of, for example, adding a new call to an existing function that expects a value to be validated but without actually validating it first, or writing `let x = Stuff { between_1_and_3: 200 };`.
